ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryTopComponent.java

```diff
@@ -125,234 +122,281 @@
+                         * property changes.
+                         */
+                        controller.regroupDisabledProperty().addListener((Observable unused) -> Platform.runLater(() -> checkForAnalyzedGroups()));
+                        controller.getGroupManager().getAnalyzedGroups().addListener((Observable unused) -> Platform.runLater(() -> checkForAnalyzedGroups()));
+
+                        /*
+                         * Dispatch a later task to call check for groups. Note
+                         * that this method displays one or more spinner(s) that
+                         * take the place of a wait cursor if there are no
+                         * analyzed groups yet, ingest is running, etc.
+                         */
+                        // RJCTODO: Is there a race condition here, since this task could be 
+                        // executed before the task to actually open the top component window?
+                        // It seems like this might be a sort of a hack and I am wondering 
+                        // why this can't be done in openWithSelectedDataSources instead.
+                        Platform.runLater(() -> checkForAnalyzedGroups());
+                    }
+
+                    /*
+                     * Kick off a background task to query the case database for
+                     * data sources. This task may queue another task for the
+                     * JavaFX thread to allow the user to select which data
+                     * sources for which to display images. Ultimately, a task
+                     * will be queued for the AWT EDT that will show the top
+                     * component window.
+                     */
+                    new Thread(new Task<Void>() {
+                        @Override
+                        protected Void call() throws Exception {
+                            synchronized (controllerLock) {
+                                /*
+                                 * If there is only one datasource or the
+                                 * grouping criterion is already set to
+                                 * something other than by path (the default),
+                                 * proceed to open this top component.
+                                 * Otherwise, do a dialog to allow the user to
+                                 * select the data sources for which images are
+                                 * to be displayed, then open the top component.
+                                 */
+                                List<DataSource> dataSources = controller.getSleuthKitCase().getDataSources();
+                                Map<DataSource, Boolean> dataSourcesWithTooManyFiles = new HashMap<>();
+                                // RJCTODO: At least some of this designation of "all data sources" with null seems uneccessary; 
+                                // in any case, the use of nulls and zeros here is 
```

Comment:
it is hacky and was done for expediency.

The 0 is just to put the "All data sources" option at the top of the list.

Reply:
Thanks for the insight, I will clean it up in the next iteration of my work on these classes.